parquet-column/src/main/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,484 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.parquet.column.values.bloom;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+
+import org.apache.parquet.bytes.*;
+import org.apache.parquet.column.Encoding;
+import org.apache.parquet.column.ParquetProperties;
+import org.apache.parquet.io.ParquetEncodingException;
+import org.apache.parquet.io.api.Binary;
+import org.apache.parquet.schema.PrimitiveType;
+
+/**
+ * Bloom Filter is a compat structure to indicate whether an item is not in set or probably in set. Bloom class is
+ * underlying class of Bloom Filter which stores a bit set represents elements set, hash strategy and bloom filter
+ * algorithm.
+ *
+ * Bloom Filter algorithm is implemented using block Bloom filters from Putze et al.'s "Cache-, Hash- and Space-Efficient Bloom
+ * Filters". The basic idea is to hash the item to a tiny Bloom Filter which size fit a single cache line or smaller.
+ * This implementation sets 8 bits in each tiny Bloom Filter. Tiny bloom filter are 32 bytes to take advantage of 32-bytes
+ * SIMD instruction.
+ *
+ */
+
+public abstract class Bloom<T extends Comparable<T>> {
+  // Hash strategy available for bloom filter.
+  public enum HASH {
+    MURMUR3_X64_128,
+  }
+
+  // Bloom filter algorithm.
+  public enum ALGORITHM {
+    BLOCK,
+  }
+
+  /**
+   * Default false positive probability value use to calculate optimal number of bits
+   * used by bloom filter.
+   */
+  public final double FPP = 0.01;
+
+  // Bloom filter data header, including number of bytes, hash strategy and algorithm.
+  public static final int BLOOM_HEADER_SIZE = 12;
+
+  // Bytes in a bucket.
+  public static final int BYTES_PER_BUCKET = 32;
+
+  // Minimum bloom filter data size.
+  public static final int MINIMUM_BLOOM_SIZE = 256;
+
+  // Hash strategy used in this bloom filter.
+  public HASH bloomFilterHash = HASH.MURMUR3_X64_128;
+
+  // Algorithm applied of this bloom filter.
+  public ALGORITHM bloomFilterAlgorithm = ALGORITHM.BLOCK;
+
+  private HashSet<Long> elements = new HashSet<>();
+
+  private byte[] bitset;
+  private int numBytes;
+
+  private int mask[] = new int[8];
+  private List<BytesInput> inputs= new ArrayList<>(4);
+
+  /**
+   * Constructor of bloom filter, if numBytes is zero, bloom filter bitset
+   * will be created lazily and the number of bytes will be calculated through
+   * distinct values in cache.
+   * @param numBytes The number of bytes for bloom filter bitset, set to zero can
+   *                 let it calculate number automatically by using default FPP.
+   * @param hash The hash strategy bloom filter apply.
+   * @param algorithm The algorithm of bloom filter.
+   */
+  public Bloom(int numBytes, HASH hash, ALGORITHM algorithm) {
+    if (numBytes != 0) {
+      initBitset(numBytes);
```

Comment:
what do you mean "elements = null "? clear the item or set reference to null ?
This should be already empty since it is a constructor.

Reply:
Set it to null. Better: don't give it an initializer outside of a constructor, since we won't know until a constructor runs if it should ever be created.